loki/transform/transform_extract_contained_procedures.py

```diff
@@ -0,0 +1,198 @@
+
+    See also the "driver" function ``extract_contained_procedures``, which applies this function to each
+    contained procedure of a parent procedure and additionally empties the CONTAINS section of subroutines.
+    """
+    inner = procedure.subroutine_map[name] # Fetch the subprocedure to extract (or crash with 'KeyError').
+
+    # Check if there are variables that don't have a scope. This means that they are not defined anywhere
+    # and execution cannot continue.
+    undefined = tuple(v for v in FindVariables().visit(inner.body) if not v.scope)
+    if undefined:
+        msg = f"The following variables appearing in the contained procedure '{inner.name}' are undefined "
+        msg += f"in both '{inner.name}' and the parent procedure '{procedure.name}': "
+        for u in undefined:
+            msg += f"{u.name}, "
+        raise RuntimeError(msg)
+
+    ## PRODUCING VARIABLES TO INTRODUCE AS DUMMY ARGUMENTS TO `inner`.
+    # Produce a list of variables defined in the scope of `procedure` that need to be resolved in `inner`'s scope
+    # by introducing them as dummy arguments.
+    # The second line drops any derived type fields, don't want them, since want to resolve the derived type itself.
+    vars_to_resolve = [v for v in FindVariables().visit(inner.body) if v.scope == procedure]
+    vars_to_resolve = [v for v in vars_to_resolve if not v.parent]
+
+    # Save any `DeferredTypeSymbol`s for later, they are in fact defined through imports in `procedure`,
+    # and therefore not to be added as arguments to `inner`. (the next step removes them from `vars_to_resolve`)
+    var_imports_to_add = tuple(v for v in vars_to_resolve if isinstance(v, DeferredTypeSymbol))
+
+    # Lookup the definition of the variables in `vars_to_resolve` from the scope of `procedure`.
+    # This provides maximal information on them.
+    vars_to_resolve = [procedure.variable_map[v.name] for v in vars_to_resolve \
+        if v.name in procedure.variable_map.keys()]
+
+    # For each array in `vars_to_resolve`, append any non-literal shape variables to `vars_to_resolve`,
+    # if not already there.
+    arr_shapes = []
+    for shape in (v.shape for v in vars_to_resolve if isinstance(v, Array)):
+        for s in shape:
+            if not isinstance(s, IntLiteral):
+                arr_shapes.append(s)
+    for v in arr_shapes:
+        if not v.name in (var.name for var in vars_to_resolve):
+            vars_to_resolve.append(v)
+
+    # 1. Rescope `vars_to_resolve` into scope of `inner`.
+    # 2. Set the intent of `vars_to_resolve` to "inout" unless specified otherwise in `procedure` scope.
+    vars_to_resolve = [v.rescope(inner) for v in vars_to_resolve]
+    newtypes = map(lambda x: x.type if x.type.intent else x.type.clone(intent = "inout"), vars_to_resolve)
+    vars_to_resolve = tuple(v.clone(type = typ) for v, typ in zip(vars_to_resolve, newtypes))
```

Comment:
I think this could be removed if you modify the arguments update further down as:
```python
    inner.arguments += tuple(
        v.clone(type=v.type.clone(intent=v.type.intent or 'inout'), scope=inner)
        for v in vars_to_resolve
    )
```

Reply:
Yes, using `or` is more handy here, thanks.

As a side note, I wish `intent` wasn't a subtype of `var.type`, but rather `var` itself, which I think would make sense and allow for `v.clone(intent = ...)`.